services/QuillLMS/app/services/analytics/segment_analytics.rb

```diff
@@ -71,12 +71,36 @@ def track_activity_pack_assignment(teacher_id, unit_id)
     })
   end
 
-  def track_activity_completion(user, student_id, activity)
+  def track_activity_completion(user, student_id, activity, activity_session)
     track({
       user_id: user&.id,
       event: SegmentIo::BackgroundEvents::ACTIVITY_COMPLETION,
       properties: activity.segment_activity.content_params.merge({student_id: student_id})
     })
+    track_activity_pack_completion(user, student_id, activity_session) if activity_pack_completed?(student_id, activity_session)
```

Comment:
`unless` might offer cleaner logic due to existing finders:

```
 track_activity_pack_completion(user, student_id, activity_session) unless activity_pack_incomplete?(student_id, activity_session)
end

def activity_pack_incomplete?(student_id, activity_session)
    ClassroomUnit
      .find_by(id: activity_session.classroom_unit_id)
      .activity_sessions
      .incomplete
      .exists?(user_id: student_id)
  end
```

Reply:
I think that, since there are not activity sessions generated for unstarted activity sessions, that solution would lead to false negatives (for instance, if a student completed the first of three activities in a pack but didn't start the other two, I think the `activity_pack_incomplete?` method would return `false`.

Reply:
@emilia-friedberg That's good to know.  

Well, in that case, you could utilize `all?` and `exists?` to clean up the block:

```
 def activity_pack_completed?(student_id, activity_session)
    classroom_unit = ClassroomUnit.find_by(id: activity_session.classroom_unit_id)
    activity_ids = classroom_unit&.unit_activities&.pluck(:activity_id)

    activity_ids.all? do |activity_id|
      ActivitySession.exists?(
        activity_id: activity_id, 
        classroom_unit: classroom_unit,
        state: ActivitySession::STATE_FINISHED,
        user_id: student_id, 
      )
  end
```

Reply:
thanks for the suggestions, y'all! I'll make this refactor